### PR TITLE
qemu.tests: Add new case of live snapshot under stress.

### PR DESCRIPTION
**Pull request by CongLi**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1321833 1321838

Based on:
https://github.com/autotest/tp-qemu/pull/701
https://github.com/avocado-framework/avocado-vt/pull/628

### CODE REVIEW

**Comment by CongLi** [comment]
20:52:52 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.with_stress
21:07:49 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.live_snapshot.with_stress

**Comment by CongLi** [comment]
11:12:53 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.with_payload.stress
11:14:27 INFO | PASS 2-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.with_payload.stop_cont

File: qemu/tests/live_snapshot_stress.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        pkg = utils.unmap_url_cache(self.test.tmpdir, link, md5sum)
+        self.vm.copy_files_to(pkg, tmp_dir)
+        logging.info("Install app: %s" % install_cmd)
+        s, o = session.cmd_status_output(install_cmd, timeout=300)
+        if s != 0:
+            raise error.TestError("Fail to install stress app(%s)" % o)
+        logging.info("Configure app: %s" % config_cmd)
+        s, o = session.cmd_status_output(config_cmd, timeout=300)
+        if s != 0:
+            raise error.TestError("Fail to conifg stress app(%s)" % o)
```
Comment (XueqiangWei) [inline qemu/tests/live_snapshot_stress.py]: As I know the stress app can be installed in framwork:
e.g.
in cfg:
bg_stress_test = autotest_control                           
test_control_file = stress.control
bg_stress_test = params.get("bg_stress_test")
check_running_cmd = params.get("check_running_cmd")
in test:
bg = utils.InterruptedThread(utils_test.run_virt_sub_test,
                                           args=(test, params, env,),
                                           kwargs={"sub_type": bg_stress_test})
bg.start()

Is it necessary to write a new function "install_stress_app" ?

Comment (CongLi) [inline qemu/tests/live_snapshot_stress.py]: Yes, for linux is better, but it does not work for Windows. :(
Current one can cover both Linux and Windows guests.

Thanks.

File: qemu/tests/cfg/live_snapshot.cfg [hunk 2]
```diff
@@ -27,3 +27,20 @@
             dd_cmd = "dd if=/dev/zero of=%s bs=4K count=1024 oflag=dsync"
             dd_timeout = 600
             copy_file_timeout = 600
+        - with_payload:
+            type = live_snapshot_stress
+            variants:
+                - stress:
```
Comment (xutian) [inline qemu/tests/cfg/live_snapshot.cfg]: why not use autotest.stress test directly?

Comment (CongLi) [inline qemu/tests/cfg/live_snapshot.cfg]: As it does not support windows guest.

Comment (xutian) [inline qemu/tests/cfg/live_snapshot.cfg]: For windows, please use win_heavyload test.

**Comment by CongLi** [comment]
14:46:23 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.live_snapshot.with_payload.stress
14:35:21 INFO | PASS 1-smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.with_payload.stress

**Comment by CongLi** [comment]
@xutian @XueqiangWei 

I have updated the case with win_heavyload and autotest.stress.
Could you help review it again?

Thanks.

File: qemu/tests/cfg/live_snapshot.cfg [hunk 3]
```diff
@@ -42,3 +42,28 @@
...
+                    Windows:
+                        timeout = 600
+                        autostress = yes
+                        stress_test = win_heavyload
+                        install_path = "C:\Program Files (x86)\JAM Software\HeavyLoad"
+                        config_cmd = 'setx -m path "%PATH%;${install_path};"'
+                        install_cmd = "start /wait DRIVE:\HeavyLoadSetup.exe /verysilent"
+                        check_cmd = 'tasklist | findstr /I  "heavyload.exe"'
+                        stop_cmd = "taskkill /T /F /IM heavyload.exe"
+                    x86_64:
```
Comment (xutian) [inline qemu/tests/cfg/live_snapshot.cfg]: "x86_64" and "i386" should belong to "Windows" section

File: qemu/tests/live_snapshot_stress.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            session.sendline(cmd)
+            if not self.app_runing():
+                return True
+            return False
+
+        error.context("stop stress app in guest", logging.info)
+        utils_misc.wait_for(_unload_stress, first=2.0,
+                            text="wait stress app quit", step=1.0, timeout=120)
+
+    def app_runing(self):
```
Comment (xutian) [inline qemu/tests/live_snapshot_stress.py]: stress_process_running  looks more good.

File: qemu/tests/live_snapshot_stress.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        """
+        error.context("launch stress app in guest", logging.info)
+        args = (self.test, self.params, self.env, self.params["stress_test"])
+        bg_test = utils_test.BackgroundTest(utils_test.run_virt_sub_test, args)
+        bg_test.start()
+        if not utils_misc.wait_for(bg_test.is_alive, first=10, step=3, timeout=100):
+            raise error.TestFail("background test start failed")
+        if not utils_misc.wait_for(self.app_runing, timeout=1800, step=5):
+            raise error.TestFail("stress app isn't running")
+        return
```
Comment (xutian) [inline qemu/tests/live_snapshot_stress.py]: unnecessary return

File: qemu/tests/live_snapshot_stress.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    def unload_stress(self):
+        """
+        stop stress app
+        """
+        def _unload_stress():
+            session = self.get_session()
+            cmd = self.params.get("stop_cmd")
+            session.sendline(cmd)
+            if not self.app_runing():
```
Comment (xutian) [inline qemu/tests/live_snapshot_stress.py]: return not self.app_running()

File: qemu/tests/cfg/live_snapshot.cfg [hunk 7]
```diff
@@ -42,3 +42,28 @@
             dd_cmd = "dd if=/dev/zero of=%s bs=4K count=1024 oflag=dsync"
```
Comment (vivianQizhu) [inline qemu/tests/cfg/live_snapshot.cfg]: Does this works for windows?

Comment (CongLi) [inline qemu/tests/cfg/live_snapshot.cfg]: @vivianQizhu 

This is another case, I have not updated it.

Comment (vivianQizhu) [inline qemu/tests/cfg/live_snapshot.cfg]: @CongLi , I guess better to add no windows here first, since yfu plan to add this to acceptance loop.

Comment (CongLi) [inline qemu/tests/cfg/live_snapshot.cfg]: This is an old case which has already been in the acceptance loop.
And checked latest acceptance test results, test pass for windows guest.
15:07:05 INFO | PASS 076-smp_4.4096m.repeat1.Host_RHEL.m6.u8.spice.qcow2.virtio_scsi.up.virtio_net.Win10.x86_64.io-github-autotest-qemu.live_snapshot.base

**Comment by xutian** [comment]
LGTM,

Thanks,
Xu

**Comment by vivianQizhu** [comment]
12:51:48 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py:420
12:51:48 ERROR| Traceback (most recent call last):
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 196, in runTest
12:51:48 ERROR|     self._runTest()
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 337, in _runTest
12:51:48 ERROR|     run_func(self, params, env)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
12:51:48 ERROR|     return fn(*args, **kwargs)
12:51:48 ERROR|   File "/home/tp-qemu/qemu/tests/live_snapshot_stress.py", line 68, in run
12:51:48 ERROR|     stress_test.action_before_start()
12:51:48 ERROR|   File "/home/tp-qemu/qemu/tests/block_copy.py", line 253, in action_before_start
12:51:48 ERROR|     return self.do_steps("before_start")
12:51:48 ERROR|   File "/home/tp-qemu/qemu/tests/block_copy.py", line 109, in do_steps
12:51:48 ERROR|     fun()
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/autotest/client/shared/error.py", line 140, in new_fn
12:51:48 ERROR|     return fn(*args, **kwargs)
12:51:48 ERROR|   File "/home/tp-qemu/qemu/tests/live_snapshot_stress.py", line 25, in load_stress
12:51:48 ERROR|     if not utils_misc.wait_for(self.stress_app_running, timeout=1800, step=5):
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/utils_misc.py", line 609, in wait_for
12:51:48 ERROR|     output = func()
12:51:48 ERROR|   File "/home/tp-qemu/qemu/tests/live_snapshot_stress.py", line 46, in stress_app_running
12:51:48 ERROR|     session = self.get_session()
12:51:48 ERROR|   File "/home/tp-qemu/qemu/tests/block_copy.py", line 85, in get_session
12:51:48 ERROR|     session = self.vm.wait_for_login(timeout=params["login_timeout"])
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/virt_vm.py", line 1030, in wait_for_login
12:51:48 ERROR|     username, password)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
12:51:48 ERROR|     return fn(*args, **kwargs)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/virt_vm.py", line 944, in login
12:51:48 ERROR|     interface=neigh_attach_if)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/remote.py", line 254, in remote_login
12:51:48 ERROR|     handle_prompts(session, username, password, prompt, timeout)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/virttest/remote.py", line 141, in handle_prompts
12:51:48 ERROR|     timeout=timeout, internal_timeout=0.5)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 853, in read_until_last_line_matches
12:51:48 ERROR|     print_func)
12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/aexpect/client.py", line 767, in read_until_output_matches
12:51:48 ERROR|     max(0, end_time - time.time()))
12:51:48 ERROR| ValueError: filedescriptor out of range in select()
12:51:48 ERROR| 
12:51:48 ERROR| Traceback (most recent call last):

12:51:48 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-38.0-py2.7.egg/avocado_vt/test.py", line 434, in _run_avocado
    raise test_exception

12:51:48 ERROR| ValueError: filedescriptor out of range in select()

12:51:48 ERROR| ERROR 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.my-provider.live_snapshot.with_payload.stress -> ValueError: filedescriptor out of range in select()

**Comment by CongLi** [comment]
@vivianQizhu 

set a low stress for test to let guest can response during stress.
Could you help test it again?

Thanks. :)

**Comment by vivianQizhu** [comment]
13:12:26 INFO | PASS 1-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.my-provider.live_snapshot.with_payload.stress

**Comment by vivianQizhu** [comment]
Acked-by: Qianqian Zhu qizhu@redhat.com

**Comment by CongLi** [comment]
Thanks @xutian @vivianQizhu , merged.

**Pull request merged**